src/puppetlabs/trapperkeeper/services/scheduler/scheduler_core.clj

```diff
@@ -19,46 +27,85 @@
-  [n f pool]
-  (let [job (wrap-with-error-logging f)]
-    (-> (at-at/interspaced n job pool)
-        :id  ; return only the ID; do not leak the at-at RecurringJob instance
-        )))
+  [n f ^Scheduler scheduler group-name]
+  (try
+    (let [job-name (Key/createUniqueName group-name)
+           job (-> (build-executable-job f job-name group-name {:interval n}))
+           schedule (-> (SimpleScheduleBuilder/simpleSchedule))
+           trigger (-> (TriggerBuilder/newTrigger)
+                       (.withSchedule schedule)
+                       (.startNow)
+                       (.build))]
+      (.scheduleJob scheduler job trigger)
+      (.getJobKey trigger))
+    (catch SchedulerException e
+      ; this can occur if the interface is being used while the scheduler is shutdown
+      (log/error e (i18n/trs "Failed to schedule job")))))
 
 (defn after
-  ; See docs on the service protocol,
-  ; puppetlabs.enterprise.services.protocols.scheduler
-  [n f pool]
-  (let [job (wrap-with-error-logging f)]
-    (-> (at-at/after n job pool)
-        :id  ; return only the ID; do not leak the at-at RecurringJob instance
-        )))
+  [n f ^Scheduler scheduler group-name]
+  (try
+    (let [job-name (Key/createUniqueName group-name)
+          job (build-executable-job f job-name group-name)
+          trigger (-> (TriggerBuilder/newTrigger)
+                      (.startAt (DateBuilder/futureDate n DateBuilder$IntervalUnit/MILLISECOND))
+                      (.build))]
+      (.scheduleJob scheduler job trigger)
+      (.getJobKey trigger))
+    (catch SchedulerException e
+      ; this can occur if the interface is being used while the scheduler is shutdown
+      (log/error e (i18n/trs "Failed to schedule job")))))
+
 
 (defn stop-job
-  "Gracefully stops the job specified by 'id'."
-  [id pool]
-  (at-at/stop id pool))
+  "Returns true, if the job was deleted, and false if the job wasn't found."
+  [^JobKey id ^Scheduler scheduler]
+  (try
+    (.deleteJob scheduler id)
+    (catch SchedulerException e
+      ; this can occur if the interface is being used while the scheduler is shutdown
+      (log/debug e (i18n/trs "Failure stopping job"))
+      false)))
+
+(defn get-all-jobs
+  [^Scheduler scheduler]
+  (try
+    (let [groups (seq (.getJobGroupNames scheduler))
+          extract-keys (fn [group-name] (seq (.getJobKeys scheduler (GroupMatcher/jobGroupEquals group-name))))]
+      (mapcat extract-keys groups))
+    (catch SchedulerException e
+      ; this can occur if the interface is being used while the scheduler is shutdown
+      (log/debug e (i18n/trs "Failure getting all jobs"))
+      [])))
 
 (defn stop-all-jobs!
-  "Stops all of the specified jobs."
-  [jobs pool]
-  (doseq [job jobs]
-    (stop-job job pool))
+  [^Scheduler scheduler]
+  (when-not (.isShutdown scheduler)
+    (try
+      (doseq [job (get-all-jobs scheduler)]
+        (.deleteJob scheduler job))
+      (.shutdown scheduler true)
```

Comment:
This will wait until jobs have completed, right? But if there's an Exception raised when calling delete on any job we'll skip this?

We might want to catch SchedulerException in the doseq loop and log a message about failing to stop job <foo>. And then call shutdown in a timeout and if the timeout elapses tell it to shutdown w/o waiting? I don't know about the reliability of shutting down Quartz, but I know we've done similar to the timeout thing in file-sync client.

Reply:
Good point.  I'll fix it.